File: BackEnd/src/main/java/hector/vico/EntreBicis/entity/Usuari.java
package hector.vico.EntreBicis.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDate;

import org.springframework.format.annotation.DateTimeFormat;

/**
 * Entitat que representa un usuari dins del sistema EntreBicis.
 * 
 * <p>Els usuaris poden tenir diferents rols (ADMIN o CICLISTA) i estan associats a rutes i recompenses.</p>
 * 
 * Es guarda a la taula <b>usuari</b> de la base de dades.
 */
@Entity
@Data  // Lombok: genera getters, setters, toString, equals i hashCode
@Table(name = "usuari")
@NoArgsConstructor  // Constructor buit per a JPA
@AllArgsConstructor // Constructor amb tots els camps
public class Usuari {

    @Id
    private String email;  // Clau primària, identificador únic de l’usuari

    private String nom;  // Nom de l'usuari

    private String cognoms;  // Cognoms de l'usuari

    @DateTimeFormat(pattern = "yyyy-MM-dd")
    private LocalDate dataNaixement;  // Data de naixement de l’usuari

    private String poblacio;  // Població o ciutat de residència

    private String telefon;  // Número de telèfon de contacte

    private String contrasenya;  // Contrasenya xifrada per accedir al sistema

    @DateTimeFormat(pattern = "yyyy-MM-dd")
    private LocalDate dataAlta;  // Data en què l’usuari es va registrar

    private int saldoPunts;  // Punts acumulats per l’usuari (serveixen per recompenses)

    @Lob  // Indica que aquest camp pot ser gran (foto guardada com a byte[])
    private byte[] foto;  // Foto de perfil de l'usuari

    @Enumerated(EnumType.STRING)
    private PermisUsuari rol;  // Rol de l’usuari: ADMIN o CICLISTA

    @Column(length = 255)
    private String observacions;  // Notes o comentaris addicionals sobre l’usuari
}

File: FrontEnd/app/src/main/java/cat/copernic/hvico/entrebicis/Core/Model/Usuari.kt
package cat.copernic.hvico.entrebicis.Core.Model

/**
 * Representa un usuari del sistema EntreBicis, ja sigui ciclista o administrador.
 *
 * @property contrasenya Contrasenya codificada de l'usuari.
 * @property nom Nom de l'usuari.
 * @property cognoms Cognoms de l'usuari.
 * @property email Correu electrònic de l'usuari (actua com a identificador principal).
 * @property dataNaixement Data de naixement de l'usuari (format ISO: yyyy-MM-dd).
 * @property telefon Número de telèfon de l'usuari (ha de tenir 9 dígits).
 * @property poblacio Ciutat o poble de residència de l'usuari.
 * @property dataAlta Data d'alta o registre al sistema (format ISO: yyyy-MM-dd).
 * @property rol Rol de l'usuari (CICLISTA o ADMIN).
 * @property saldoPunts Quantitat actual de punts disponibles a l'usuari.
 * @property foto Ruta o URL de la foto de perfil (opcional).
 */
data class Usuari(
    val contrasenya: String,
    val nom: String,
    val cognoms: String,
    val email: String,
    val dataNaixement: String,
    val telefon: String,
    val poblacio: String,
    val dataAlta: String,
    val rol: PermisUsuari,
    val saldoPunts: Int?,
    val foto: String? = null
)